### PR TITLE
feat(auth): add --no-browser flag for browser-less OAuth login

### PR DESCRIPTION
**Pull request by timothedelion**
## Summary

Adds `ggshield auth login --no-browser` for environments without a browser (SSH sessions, headless servers).

When the flag is set, ggshield:
1. Builds the authorize URL with `redirect_uri=urn:ietf:wg:oauth:2.0:oob` (the standard OAuth out-of-band sentinel).
2. Prints the URL to stderr — the user opens it on another device.
3. Reads the displayed code from stdin via `click.prompt`.
4. Exchanges it at `/v1/oauth/token` with the same OOB redirect_uri and the existing `code_verifier`.

PKCE still binds the exchange to the local CLI. The localhost listener is not started in OOB mode. State validation is intentionally absent (no redirect to inspect) — documented inline.

Server side: MR in GIM : https://gitlab.gitguardian.ovh/gg-code/prm/ward-runs-app/-/merge_requests/24425

image
image
image

Loom video : https://www.loom.com/share/4b358456740b4905a16ee9d2af2a8c6f

## Test plan

- [x] New unit tests cover: URL contains the sentinel + no browser opened + no listener started, POST body has the sentinel, code-stripping, empty-code error, 4xx surfacing, and `--no-browser` rejection with `--method=token`.
- [x] All pre-existing auth tests still pass (358/358).
- [x] `black --check` / `isort --check` clean on changed files.
- [ ] Manual smoke against a backend with OOB support.

Refs: SI-2994

### CODE REVIEW

File: ggshield/verticals/auth/oauth.py [hunk 1]
```diff
@@ -104,11 +139,65 @@ def oauth_process(
...
+            raise
+        if not authorization_code:
+            raise UnexpectedError("No authorization code was provided.")
+        click.echo(f"Authorization code: {_mask_code(authorization_code)}", err=True)
+
+        try:
+            self._claim_token(authorization_code)
+        except OAuthError as error:
+            raise UnexpectedError(error.message)
+        token_data = self._validate_access_token()
```
Comment (6d7a) [inline ggshield/verticals/auth/oauth.py]: `self._validate_access_token()` may raise an `OAuthError` as well, so we should move it into the `try` block.

**Review by 6d7a — CHANGES_REQUESTED** [review]
~Thank you for the PR and sorry for the long wait. The changes look solid, I only have one  change request.
I have one concern though about the general scope of this: We already have `ggshield auth login --method token` which provides a completely headless auth flow. AFAIK OAuth 2.0 OOB is a legacy out-of-band flow, so if we want to adopt a standard for replacing our custom headless auth flow, shouldn't we go with a future-proof protocol?~
Saw the discussion on the sibling MR 😅

**Review by 6d7a — CHANGES_REQUESTED** [review]
Thank you for the PR and sorry for the long wait. The changes look solid, I only have one general suggestion: We already have the `--method` CLI parameter, could you wire up the OOB flow to it so that we can use it like `ggshield auth login --method oob`. That would be more consistent with the existing interface.

File: tests/unit/cmd/auth/test_login.py [hunk 2]
```diff
@@ -1010,3 +1010,201 @@ def test_api_instance_url(self, instance_url, cli_fs_runner, monkeypatch):
         exit_code, output = self.run_cmd(cli_fs_runner)
         assert exit_code == ExitCode.SUCCESS, output
         self._webbrowser_open_mock.assert_called()
+
+
+OOB_REDIRECT_URI = "urn:ietf:wg:oauth:2.0:oob"
```
Comment (HelifeWasTaken) [inline tests/unit/cmd/auth/test_login.py]: `OOB_REDIRECT_URI` is already exported from `ggshield.verticals.auth.oauth`.
Redefining it here creates a silent drift risk — if the value ever changes in
the source, this copy stays wrong while tests keep passing.
Add `from ggshield.verticals.auth.oauth import OOB_REDIRECT_URI` to the
import block at the top of the file and remove this local definition.

File: ggshield/verticals/auth/oauth.py [hunk 3]
```diff
@@ -36,6 +43,20 @@
...
 logger = logging.getLogger(__name__)
 
+# Number of leading characters of an authorization code we keep visible when
+# echoing it back; the rest is replaced by `*` so the full code never lands
+# in the terminal scrollback in plain text.
+_OOB_CODE_VISIBLE_PREFIX = 4
+
+
+def _mask_code(code: str) -> str:
+    """Return ``code`` with all but the first 4 characters replaced by ``*``."""
```
Comment (HelifeWasTaken) [inline ggshield/verticals/auth/oauth.py]:
```suggestion
"""Return ``code`` with all but the first 4 characters replaced by ``*`` (fully masked if ≤ 4 chars)."""
```
The docstring says "all but the first 4 characters replaced by `*`", but the
guard `<= _OOB_CODE_VISIBLE_PREFIX` means codes whose length is exactly 4
(or shorter) have ALL characters masked — not just the suffix. The docstring
should reflect that edge case.

File: tests/unit/cmd/auth/test_login.py [hunk 4]
```diff
@@ -1010,3 +1010,201 @@ def test_api_instance_url(self, instance_url, cli_fs_runner, monkeypatch):
...
+        assert "Success! You are now authenticated" in result.output
+        # Stderr should mention pasting the code.
+        assert (
+            "paste" in result.output.lower() or "paste" in (result.stderr or "").lower()
+        )
+
+        # Pasted code must not appear in plain text in the terminal output.
+        # Only the first 4 chars are echoed back, followed by `*` placeholders.
+        assert "some_authorization_code" not in result.output
+        assert "some" + "*" * (len("some_authorization_code") - 4) in result.output
```
Comment (HelifeWasTaken) [inline tests/unit/cmd/auth/test_login.py]:
```suggestion
assert (
    "some" + "*" * (len("some_authorization_code") - 4) in result.output
    or "some" + "*" * (len("some_authorization_code") - 4)
    in (result.stderr or "")
)
```
This assertion only checks `result.output`, relying on `CliRunner`'s default
`mix_stderr=True` to include the stderr echo in output. The "paste" check
just above already defensively handles both `mix_stderr` settings with an `or`
fallback. Apply the same pattern here for consistency.

**Comment by HelifeWasTaken** [comment]
Just nitpicks I could find because I was doing a review not really security issues if @6d7a concerns are addressed

**Review by 6d7a — APPROVED** [review]
Thank you for the changes!
